### ISSUE TITLE
Support YAML syntax highlighting

### ISSUE THREAD

**Issue by samuelba** [doxygen/doxygen#7551]
Would be nice if yaml syntax highlighting in markdown code blocks is supported.

**Comment by albert-github** [comment]
- What do you mean by `arkdown code blocks`? are these the parts in fenceded code blocks (`~~~`) ?
- Could you give a reference how this would look like. 
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce this enhancement request / provblem?  Please don't add external links as they might not be persistent.

**Comment by samuelba** [comment]
Yes, I mean fenced code blocks. The syntax highlighting is already working e.g. for `~~~cpp` or `~~~xml` but not for `~~~yaml`.

**Comment by albert-github** [comment]
I think in this case the "problem" is that YAML syntax is not supported by doxygen. 
For cpp it is clear here the doxygen code coloring is used (like it is for other languages as supported by doxygen).

Can you also give a small example of a XML fenced code block? it might be that this some side effect  from something else and only some words are highlighted.

- so would be nice to have 2 examples:
  - XML 
  - YAML
- in which doxygen version did you see the XML code coloring?

**Comment by samuelba** [comment]
I use version 1.8.13. I don't think the xml coloring is a side effect, it looks pretty good.

Build example:
```
mkdir build && cd build
cmake ..
```

Then open `doc/html/index.html`.

[example_doc.zip](https://github.com/doxygen/doxygen/files/4134172/example_doc.zip)

Output:

![image](https://user-images.githubusercontent.com/3610764/73454437-651bb680-436e-11ea-8f58-a0cb38b8c4d4.png)

**Comment by samuelba** [comment]
Result of the yaml highlighting could be something like this:

```yaml
# Comment
map: # Another comment
  sequence:
    - name: "a"
    - name: "b"
    - name: "c"
  value_1: 1
  value_2: 2
  value_3: 3
```

**Comment by albert-github** [comment]
For XML there is for input only a code lexer (in xmlcode.h: " Only support syntax highlighting of code at the moment."). to implement the code coloring for YAML a new code parser would probably be needed.

**Comment by aotto1968** [comment]
There is already a good code parser I use for syntax highlight of __NON__ doxygen supported languages, it is called `source-highlight`

I arrange it to be used like `snippet` with output in `html`, and an `aliases` for import into *doxygen* documentation

```
ALIASES += "htmlcode{4}=\b Example from <tt> \1\fileextension </tt> \b → \4  ^^ <div class=\"htmlcode\"> \htmlinclude \1_\3.\2.html </div>"
```

In my `makefile ` I use

```
$(top_srcdir)/sbin/source-highlight.bash --dir $(abs_builddir)/docs --regex ErrorNoRaise_example $(abs_srcdir)/*/server.*
```

The wrapper is quite simple
```
## usage …
##
## > used in directory "example" to create a "html-output-file" in "docs" directory 
## > from the code between "server_example"
##
## source-highlight.bash --dir docs --regex server_example vb/MyServer.vb

error() {
  echo "ERROR: $@" 1>&2
  [[ -r "$TARGET_FILE" ]] && rm "$TARGET_FILE"
  exit 1
}

usage() {
  error "source-highlight.bash --dir TARGET-DIR --regex REGEX-RANGE SOURCE-FILE..."
}

TARGET_DIR="."
REGEX_RANGE=""
TARGET_FILE=""

while (( $# > 0 )) ; do
  case "$1" in
    --stdout) TARGET_FILE="stdout"
              shift 1
    ;;
    --dir)    TARGET_DIR="$2"
              shift 2
    ;;
    --regex)  REGEX_RANGE="$2"
              shift 2
    ;;
    -*)       usage ;;
    *)        break ;;
  esac
done

test ! -d "$TARGET_DIR" && mkdir -p "$TARGET_DIR"

if [[ "$REGEX_RANGE" != "" ]] ; then
  REGEX_RANGE_CMD="--regex-range \\[$REGEX_RANGE\\]"
else
  REGEX_RANGE_CMD=""
fi

for SOURCE_FILE ; do

  EXTENSION="${SOURCE_FILE##*.}"
  if [[ "$TARGET_FILE" != "stdout" ]] ; then
    TARGET_FILE="${SOURCE_FILE##*/}"
    TARGET_FILE="$TARGET_DIR/${TARGET_FILE%.*}${REGEX_RANGE:+_}${REGEX_RANGE}.$EXTENSION.html"
    TARGET_CMD="--output $TARGET_FILE"
  else
    TARGET_CMD=""
  fi

  test -r "$SOURCE_FILE" || error "SOURCE_FILE='$SOURCE_FILE' not readable"

  [[ "$TARGET_FILE" != "stdout" && -r "$TARGET_FILE" && "$TARGET_FILE" -nt "$SOURCE_FILE" ]] && continue

  echo "work on… $SOURCE_FILE"

  if [[ ! -z "$REGEX_RANGE" ]] ; then
    (( $(grep --fixed-strings --count "[$REGEX_RANGE]" $SOURCE_FILE) != 2 )) && \
      error "range '$REGEX_RANGE' was not found '2' times in '$SOURCE_FILE'"
  fi

  source-highlight $REGEX_RANGE_CMD --data-dir="$NHI1_HOME/etc/lang" --out-format=html-css --input "$SOURCE_FILE" $TARGET_CMD ||
    error "TARGET_FILE='$TARGET_FILE' not createable"

done

exit 0
```

**Comment by enerjazzer** [comment]
great suggestion about 
\code{filter:my/path/to/filter.bash}
Should be work for \snippet too

But it can be bothersome to write it in every place if there are many places like these, so I think we also need an additional global setting for custom highlighters per extension, specified in the doxygen config file, like:
SYNTAX_HIGHLIGHTERS+=".yaml:my/path/to/filter.bash"
SYNTAX_HIGHLIGHTERS+=".toml:my/path/to/toml_highlighter"
SYNTAX_HIGHLIGHTERS+=".x|.y|.z:my/path/to/xyz_joint_highlighter"

**Comment by vimarshacooray** [comment]
+1

**Comment by ischoegl** [comment]
+1

**Comment by ischoegl** [comment]
Fwiw, the [`pygments`](https://pygments.org/docs/quickstart/) package provides another lexer with support for an even wider range of languages (see [here](https://pygments.org/languages/)). For files, conversion to stdout is a simple as 

```
$ pygmentize -f html something.yml
```

Myself, I was not able to make this work with an `ALIASES` definition, as I could not find a way to pipe the argument of an ALIASES definition to a filter (if I understand correctly, @aotto1968's approach inserts preformatted html code generated from stand-alone files in a separate preprocessing step instead).

Having a section of fenced code
```
\code{.yaml}
# Comment
map: # Another comment
  sequence:
    - name: "a"
    - name: "b"
    - name: "c"
  value_1: 1
  value_2: 2
  value_3: 3
\endcode
```

highlighted using an approach along the lines of @enerjazzer suggestion that would allow for

```
SYNTAX_HIGHLIGHTERS+=".yaml:pygmentize -f html"`
```
(i.e. provide shell command plus arguments) would indeed be great! 

It is presumably still possible to create an `INPUT_FILTER` to preprocess files and stitch content together. Of course, I may be overlooking something?

**Comment by phetdam** [comment]
@ischoegl I took a pre-processing approach with Pygments where my use case was syntax highlighting fenced code blocks in standalone Markdown files for languages not known to Doxygen, in particular CMake. My project already uses Python and I like Pygments, having used it in LaTeX via the `minted` package as `listings` output looked very ugly in comparison.

In my case I wrote a Python script that would inject raw HTML output from Pygments, but only for targeted triple-backtick code blocks like the following that were scoped with specific HTML comments. For example:

~~~
<!-- input.md -->

... some text ...

<!-- pygmentize: on -->

```cmake
find_package(CURL)
```

<!-- pygmentize: off -->

... more text ...
~~~

Code blocks between the `<!-- pygmentize: (on|off) -->` tags would be replaced with Pygments HTML obtained as if with:

```python
from pygments import HtmlFormatter, get_lexer_by_name, highlight

# ...

# select lexer by the language name, e.g. "cmake", "cpp", etc.
lexer = get_lexer_by_name(lang_name)
# use HtmlFormatter without wrapping (more on this later)
# note: could use classprefix=pygments- or something if desired as well
formatter = HtmlFormatter(nowrap=True)
# generate HTML fragment with <pre class="fragment">. this is necessary for the
# HTML fragment to be formatted consistently with Doxygen HTML
html = f"<pre class=\"fragment\">{highlight(code_block, lexer, formatter)}</pre>"
print(html)

# ...
```

On the `input.md` example code shown above, the output would look like the following:

~~~
<!-- input-pygmentized.md -->

... some text ...


<!-- HTML generated with pyginject BEGIN -->
<span class="nb">find_package</span><span class="p">(</span><span class="s">CURL</span><span class="p">)</span>
<!-- HTML generated with pyginject END -->


... more text ...
~~~

So instead of including `input.md` in my Doxygen sources, I include `input-pygmentized.md` instead. For one-dark styling of the code blocks, I added to [`HTML_EXTRA_STYLESHEET`](https://www.doxygen.nl/manual/config.html#cfg_html_extra_stylesheet) the `pygments-one-dark.css` file generated by the following:

```shell
pygmentize -S one-dark -f html > pygments-one-dark.css
```

I haven't fully integrated this entire process with [`doxygen_add_docs`](https://cmake.org/cmake/help/latest/module/FindDoxygen.html#command:doxygen_add_docs) in my CMake build config but the output is promising so far. In my opinion, if one is already using Python in their project, this approach is less cumbersome than trying to massage different languages into a C-like syntax for Doxygen to consume via filter.
